Sources/Signal.swift

```diff
@@ -38,132 +56,207 @@ public final class Signal<Value, Error: Swift.Error> {
-							shouldDispose = true
+					// Check if the status has been bumped to `terminating` due to a
+					// concurrent or a recursive termination event.
+					if case .terminating = signal.state, let snapshot = signal.shouldReallyTerminate() {
+						for observer in snapshot.observers {
+							observer.action(snapshot.associated)
 						}
 
-						if isTerminating {
-							terminated = true
-							shouldDispose = true
-						}
+						shouldDispose = true
 					}
+				}
+
+				// End of the main protected section.
+				signal.sendLock.unlock()
+
+				// Check if the status has been bumped to `terminating` due to a
+				// concurrent termination event that has not been caught in the main
+				// protected section.
+				if !shouldDispose, case .terminating = signal.state {
+					signal.sendLock.lock()
 
-					sendLock.unlock()
-
-					// Based on the implicit memory order, any updates to the
-					// `interruptedState` should always be visible after `sendLock` is
-					// released. So we check it again and handle the interruption if
-					// it has not been taken over.
-					if !shouldDispose && !terminated && !isTerminating, let state = interruptedState {
-						sendLock.lock()
-
-						// `terminated` before acquring the lock could be a false negative,
-						// since it might race against other concurrent senders until the
-						// lock acquisition above succeeds. So we have to check again if the
-						// signal is really still alive.
-						if !terminated {
-							interrupt(state.observers)
-							shouldDispose = true
+					if let snapshot = signal.shouldReallyTerminate() {
+						for observer in snapshot.observers {
+							observer.action(snapshot.associated)
 						}
 
-						sendLock.unlock()
+						shouldDispose = true
 					}
 
-					if shouldDispose {
-						// Dispose only after notifying observers, so disposal
-						// logic is consistently the last thing to run.
-						signal.generatorDisposable?.dispose()
-					}
+					signal.sendLock.unlock()
+				}
+
+				if shouldDispose {
+					// Dispose only after notifying observers, so disposal
+					// logic is consistently the last thing to run.
+					signal.swapDisposable()?.dispose()
 				}
 			}
 		}
 
 		generatorDisposable = generator(observer)
 	}
 
-	deinit {
-		if state.swap(nil) != nil {
-			// As the signal can deinitialize only when it has no observers attached,
-			// only the generator disposable has to be disposed of at this point.
-			generatorDisposable?.dispose()
+	/// Return the state snapshot if the signal is in the `terminating` state, or
+	/// `nil` otherwise.
+	///
+	/// Calling this method as a result of a false positive `terminating` check
+	/// is permitted. `nil` would be returned in this case.
+	///
+	/// - note: The `updateLock` would be acquired.
+	///
+	/// - returns:
+	///   The state snapshot associated with a termination event, or `nil`.
+	private func shouldReallyTerminate() -> SignalStateSnapshot<Value, Error, Event<Value, Error>>? {
```

Comment:
This function is only used like this:

```swift
if let snapshot = signal.shouldReallyTerminate() {
    for observer in snapshot.observers {
        observer.action(snapshot.associated)
    }
}  
```

That suggests that the code to notify observers should move into this function.

It probably also makes more sense as a block in `init` since it's only used there.